File: leetcode/856_Score_of_Parentheses/Score_of_Parentheses.java
class Solution {
    public int scoreOfParentheses(String S) {
        Stack<String> pStack = new Stack<>();
        for (int i = 0; i < S.length(); i++) {
            if (S.charAt(i) == '(') pStack.push("(");
            else if (S.charAt(i) == ')') {
                if (pStack.empty()) return 0;
                else if (pStack.peek().equals("(")) {
                    pStack.pop();
                    pStack.push("1");
                }
                else {
                    int res = 0;
                    while (!pStack.empty() && !pStack.peek().equals("(")) {
                        res += Integer.parseInt(pStack.pop());
                    }
                    if (!pStack.empty()) {
                        res *= 2;
                        pStack.pop();
                    }
                    pStack.push(Integer.toString(res));
                }
            }
        }
        if (pStack.size() < 1) return 0;
        else if (pStack.size() == 1) return Integer.parseInt(pStack.peek());
        else {
            int res = 0;
            while (!pStack.empty()) res += Integer.parseInt(pStack.pop());
            return res;
        }
    }
}

File: leetcode/856_Score_of_Parentheses/Score_of_Parentheses.cpp
class Solution {
public:
    int scoreOfParentheses(string S) {
        vector<int> intStack = vector<int>();
        for (int i = 0; i < S.size(); i++) {
            if (S[i] == '(') intStack.push_back(0);
            else if (intStack.back() == 0) intStack.back() += 1;
            else {
                int tmp = 0;
                while (!intStack.empty() && intStack.back() != 0) {
                    tmp += intStack.back() * 2;
                    intStack.pop_back();
                }
                intStack.back() += tmp;
            }
        }
        int res = 0;
        while (!intStack.empty()) {
            res += intStack.back();
            intStack.pop_back();
        }
        return res;
    }
};
